### PR TITLE
Atom starter game ios fixes

### PR DESCRIPTION
**Pull request by moudgils**
- Use low end pipeline on mobile by default for BootStrapComponent (via Palification)
 - Track the need to bind null heap within Argument buffers
 - Only bind the null heap if its needed
 - Track its usage for Vertex/Fragment stages
 - Increase null dummy buffer to 1K to address GPU crash oon thee first frame

### CODE REVIEW

**Review by AMZN-tpeng — APPROVED** [review]

**Review by hershey5045 — APPROVED** [review]

File: Gems/Atom/Bootstrap/Code/Source/Platform/Android/BootstrapSystemComponent_Traits_Platform.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#define BOOTSTRAPSYSTEMCOMPONENT_PIPELINE_NAME     "passes/LowEndRenderPipeline.azasset"
```
Comment (rgba16f) [inline Gems/Atom/Bootstrap/Code/Source/Platform/Android/BootstrapSystemComponent_Traits_Platform.h:9]: I believe the naming standard is AZ_TRAIT_*

File: Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp [hunk 2]
```diff
@@ -314,13 +315,17 @@ namespace AZ
                 // Create a render pipeline from the specified asset for the window context and add the pipeline to the scene.
                 // When running with no Asset Processor (for example in release), CompileAssetSync will return AssetStatus_Unknown.
                 AzFramework::AssetSystem::AssetStatus status = AzFramework::AssetSystem::AssetStatus_Unknown;
+                default_pipeline_name = BOOTSTRAPSYSTEMCOMPONENT_PIPELINE_NAME;
```
Comment (rgba16f) [inline Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp:318]: This overrides the cvar with a define? Doesn't that defeat the reason to use a cvar?

File: Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp [hunk 3]
```diff
@@ -38,6 +38,10 @@
 #include <Atom/Bootstrap/BootstrapNotificationBus.h>
 
 #include <Atom/RPI.Reflect/System/AnyAsset.h>
+#include <AzCore/Console/IConsole.h>
+#include <BootstrapSystemComponent_Traits_Platform.h>
+
+AZ_CVAR(AZ::CVarFixedString, default_pipeline_name, "passes/MainRenderPipeline.azasset", nullptr, AZ::ConsoleFunctorFlags::DontReplicate, "Default Render pipeline name");
```
Comment (rgba16f) [inline Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp:44]: The default cvar value here should come from the az traits define

File: Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp [hunk 4]
```diff
@@ -314,13 +315,17 @@ namespace AZ
                 // Create a render pipeline from the specified asset for the window context and add the pipeline to the scene.
                 // When running with no Asset Processor (for example in release), CompileAssetSync will return AssetStatus_Unknown.
                 AzFramework::AssetSystem::AssetStatus status = AzFramework::AssetSystem::AssetStatus_Unknown;
+                default_pipeline_name = BOOTSTRAPSYSTEMCOMPONENT_PIPELINE_NAME;
+                const AZ::CVarFixedString pipelineName = static_cast<AZ::CVarFixedString>(default_pipeline_name);
```
Comment (rgba16f) [inline Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp:319]: This line seems unnecessary? default_pipeline_name is already a CVarFixedString.

File: Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp [hunk 5]
```diff
@@ -314,13 +315,17 @@ namespace AZ
...
                 AzFramework::AssetSystemRequestBus::BroadcastResult(
-                    status, &AzFramework::AssetSystemRequestBus::Events::CompileAssetSync, m_defaultPipelineAssetPath);
-                AZ_Assert(status == AzFramework::AssetSystem::AssetStatus_Compiled || status == AzFramework::AssetSystem::AssetStatus_Unknown, "Could not compile the default render pipeline at '%s'", m_defaultPipelineAssetPath.c_str());
+                    status, &AzFramework::AssetSystemRequestBus::Events::CompileAssetSync, pipelineName.data());                
+
```
Comment (rgba16f) [inline Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp:322]: added extra whitespace

File: Gems/Atom/RHI/Metal/Code/Source/RHI/ArgumentBuffer.cpp [hunk 6]
```diff
@@ -282,12 +283,14 @@ namespace AZ
...
                         RHI::Ptr<Memory> nullMtlBufferMemPtr = nullDescriptorManager.GetNullImageBuffer().GetMemory();
                         mtlTextures[bufferArrayLen] = nullMtlBufferMemPtr->GetGpuAddress<id<MTLTexture>>();
+                        m_useNullDescriptorHeap = true;
                     }
                     else
                     {
                         RHI::Ptr<Memory> nullMtlBufferMemPtr = nullDescriptorManager.GetNullBuffer().GetMemory();
                         mtlBuffers[bufferArrayLen] = nullMtlBufferMemPtr->GetGpuAddress<id<MTLBuffer>>();
                         mtlBufferOffsets[bufferArrayLen] = nullDescriptorManager.GetNullBuffer().GetOffset();
+                        m_resourceBindings[shaderInputBuffer.m_name].insert(ResourceBindingData{nullMtlBufferMemPtr, .m_bufferAccess = shaderInputBuffer.m_access});
```
Comment (rgba16f) [inline Gems/Atom/RHI/Metal/Code/Source/RHI/ArgumentBuffer.cpp:293]: nit: line length

File: Gems/Atom/RHI/Metal/Code/Source/RHI/CommandList.cpp [hunk 7]
```diff
@@ -291,6 +294,9 @@ namespace AZ
                                                             
                     if(srgVisInfo != RHI::ShaderStageMask::None)
                     {
+                        bool isNullDescHeapNeeded = compiledArgBuffer.IsNullDescHeapNeeded();
+                        bindNullDescriptorHeap |= isNullDescHeapNeeded;
```
Comment (rgba16f) [inline Gems/Atom/RHI/Metal/Code/Source/RHI/CommandList.cpp:298]: nit: why do as 2 lines?

File: Gems/Atom/RHI/Metal/Code/Source/RHI/CommandList.cpp [hunk 8]
```diff
@@ -300,7 +306,8 @@ namespace AZ
                                 mtlVertexArgBuffers[slotIndex] = argBuffer;
                                 mtlVertexArgBufferOffsets[slotIndex] = argBufferOffset;
                                 bufferVertexRegisterIdMin = AZStd::min(slotIndex, bufferVertexRegisterIdMin);
-                                bufferVertexRegisterIdMax = AZStd::max(slotIndex, bufferVertexRegisterIdMax);
+                                bufferVertexRegisterIdMax = AZStd::max(slotIndex, bufferVertexRegisterIdMax);                                
+                                mtlRenderStagesForNullDescHeap = shaderResourceGroup->IsNullHeapNeededForVertexStage(srgResourcesVisInfo) ? mtlRenderStagesForNullDescHeap | MTLRenderStageVertex : mtlRenderStagesForNullDescHeap;
```
Comment (rgba16f) [inline Gems/Atom/RHI/Metal/Code/Source/RHI/CommandList.cpp:310]: nit: long line

File: Gems/Atom/RHI/Metal/Code/Source/RHI/CommandList.cpp (hunk 7 above)
Comment (moudgils) [reply Gems/Atom/RHI/Metal/Code/Source/RHI/CommandList.cpp:298]: I use isNullDescHeapNeeded separately below

File: Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp (hunk 4 above)
Comment (moudgils) [reply Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp]: Its not. I have to cast it to that or I get compile errors.

**Review by rgba16f — APPROVED** [review]

**Pull request merged**